### PR TITLE
Allow Symfony 5 components

### PR DESCRIPTION
**Pull request by l-vo**
Allow Symfony 5 components

cc @nicolas-grekas

### CODE REVIEW

File: src/Console/ConfigurationResolver.php [hunk 1]
```diff
@@ -305,6 +305,10 @@ public function getDirectory()
     {
         if (null === $this->directory) {
             $path = $this->getCacheFile();
+            if (null === $path) {
+                return null;
```
Comment (keradus) [inline src/Console/ConfigurationResolver.php]: this explicitly violate the method return type

File: src/Event/Event.php [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *     Dariusz Rumiński <dariusz.ruminski@gmail.com>
+ *
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+namespace PhpCsFixer\Event;
+
+if (class_exists(\Symfony\Contracts\EventDispatcher\Event::class)) {
```
Comment (keradus) [inline src/Event/Event.php]: please add some comment why it's needed and when it would not be needed anymore (how to cleanup, when to cleanup)

File: src/Console/ConfigurationResolver.php (hunk 1 above)
Comment (l-vo) [reply src/Console/ConfigurationResolver.php]: Indeed... Modified, thanks :)

File: src/Event/Event.php (hunk 2 above)
Comment (l-vo) [reply src/Event/Event.php]: Done :)

File: tests/Event/EventTest.php [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+ */
+
+namespace PhpCsFixer\Tests\Event;
+
+use PhpCsFixer\Event\Event;
+use PhpCsFixer\Tests\TestCase;
+
+/**
+ * @internal
+ * @coversNothing
```
Comment (julienfalque) [inline tests/Event/EventTest.php]: Should be:
```suggestion
 * @covers \PhpCsFixer\Event\Event
```

Comment (l-vo) [reply tests/Event/EventTest.php]: Fixed, thank you :)

File: src/Event/Event.php [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+} else {
+    class BaseEvent extends \Symfony\Component\EventDispatcher\Event
+    {
+    }
+}
+
+/**
+ * @author Laurent VOULLEMIER <laurent.voullemier@gmail.com>
+ */
+class Event extends BaseEvent
```
Comment (derrabus) [inline src/Event/Event.php]: The `BaseEvent` class looks unnecessary to me. You could directly define the `Event` class within the `if`/`else` block above.

```php
if (class_exists(\Symfony\Contracts\EventDispatcher\Event::class)) {
    class Event extends \Symfony\Contracts\EventDispatcher\Event
    {
    }
} else {
    class Event extends \Symfony\Component\EventDispatcher\Event
    {
    }
}
```

Comment (l-vo) [reply src/Event/Event.php]: Indeed... Thank you

File: tests/Event/EventTest.php [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+ */
+
+namespace PhpCsFixer\Tests\Event;
+
+use PhpCsFixer\Event\Event;
+use PhpCsFixer\Tests\TestCase;
+
+/**
+ * @internal
+ * @@covers \PhpCsFixer\Event\Event
```
Comment (keradus) [inline tests/Event/EventTest.php]: `@@`

File: src/Console/ConfigurationResolver.php [hunk 6]
```diff
@@ -305,11 +305,15 @@ public function getDirectory()
     {
         if (null === $this->directory) {
             $path = $this->getCacheFile();
-            $filesystem = new Filesystem();
+            if (null === $path) {
+                $absolutePath = $this->cwd;
```
Comment (keradus) [inline src/Console/ConfigurationResolver.php]: can you elaborate on this change?
it does not look related to `Event` at first glance

**Comment by keradus** [comment]
@l-vo , please ping us when 2nd point of https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/4608#pullrequestreview-307590178 is ready

File: src/Console/ConfigurationResolver.php (hunk 6 above)
Comment (derrabus) [reply src/Console/ConfigurationResolver.php]: @keradus `Filesystem::isAbsolutePath()` does not accept `null` anymore, so that case is handled separately.

Comment (keradus) [reply src/Console/ConfigurationResolver.php]: ah. thanks for clarification !

File: tests/Console/ConfigurationResolverTest.php [hunk 7]
```diff
@@ -1143,6 +1143,47 @@ public function provideDeprecatedFixerConfiguredCases()
         ];
     }
 
+    public function provideGetDirectoryCases()
+    {
+        return [
+            [null, '/my/path/my/file', 'path/my/file'],
```
Comment (julienfalque) [inline tests/Console/ConfigurationResolverTest.php]: Shouldn't the expected relative path be `my/file` instead of `path/my/file`?

Comment (l-vo) [reply tests/Console/ConfigurationResolverTest.php]: Maybe in the case where there is no cache file,`getDirectory` doesn't behave as expected. But it's the current behavior. Fixing this potential issue is not the goal of this PR and I'm not sur this wouldn't be considered as a BC break.

**Comment by l-vo** [comment]
@keradus CI steps to test against Symfony 4 and Symfony 5 components added

File: .travis.yml [hunk 8]
```diff
@@ -78,6 +79,11 @@ jobs:
             php: 5.6
             env: COMPOSER_FLAGS="--prefer-stable --prefer-lowest"
 
+        -   <<: *STANDARD_TEST_JOB
+            stage: Test
+            php: 7.1
+            env: SYMFONY_DEPRECATIONS_HELPER=disabled PHP_CS_FIXER_TEST_USE_LEGACY_TOKENIZER=1 SYMFONY_VERSION="~3.0"
```
Comment (keradus) [inline .travis.yml]: why `SYMFONY_DEPRECATIONS_HELPER=disabled` ?
why `PHP_CS_FIXER_TEST_USE_LEGACY_TOKENIZER=1` ?

**Comment by keradus** [comment]
@l-vo , can we reuse some of existing jobs?
i would really like to avoid extra 2 jobs

File: .travis.yml (hunk 8 above)
Comment (l-vo) [reply .travis.yml]: Job not used anymore, using `COMPOSER_FLAGS="--prefer-stable --prefer-lowest"` instead which already installs Symfony 3 components

**Comment by l-vo** [comment]
@keradus PR modified to reuse existing jobs

**Comment by alexander-schranz** [comment]
@julienfalque another way would be use the `.phar` for php-coveralls as it is just a dev dependency. Example: https://github.com/sulu/SuluContentBundle/blob/756944dbb17a27ed840fc12d0d13203ccb4e7aed/.circleci/config.yml#L135-L138

**Comment by andreybolonin** [comment]
Screen Shot 2019-11-21 at 1 09 53 PM

**Comment by l-vo** [comment]
@julienfalque rebased on 2.15

**Comment by keradus** [comment]
@julienfalque , php-coveralls issue solved. can you review once more?
@l-vo , why changing the baseline to 2.15? if sth would not be compatible, we would need to fix it then on both, 2.15 and 2.16, instead of only 2.16

**Comment by julienfalque** [comment]
@keradus [I did ask to rebase onto 2.15](4608#pullrequestreview-318660553). Travis build was green on 2.16 so merging 2.15 into 2.16 should be fine anyway.

File: .travis.yml [hunk 9]
```diff
@@ -88,6 +89,7 @@ jobs:
             <<: *STANDARD_TEST_JOB
             stage: Test
             php: 7.2
+            env: SYMFONY_DEPRECATIONS_HELPER=disabled PHP_CS_FIXER_TEST_USE_LEGACY_TOKENIZER=1 SYMFONY_VERSION="5.0.x-dev" MINIMUM_STABILITY=dev
```
Comment (julienfalque) [inline .travis.yml]: Symfony 5.0 stable is available now, we can use it rather than the `-dev` version.

Comment (l-vo) [reply .travis.yml]: Indeed... Fixed :)

Comment (julienfalque) [reply .travis.yml]: Maybe `^5.0` instead of `~5.0.0`? Also the line with `$MINIMUM_STABILITY` isn't necessary anymore.

**Comment by fabpot** [comment]
:+1: Looks great! That's the only missing piece to allow me to upgrade fabbot.io to Symfony 5.0 :)

**Comment by TomasVotruba** [comment]
Bloker for https://github.com/getsentry/sentry-php/pull/925

**Comment by julienfalque** [comment]
Thank you @l-vo.

File: .composer-require-checker.json [hunk 10]
```diff
@@ -1,6 +1,8 @@
 {
     "symbol-whitelist" : [
+        "Symfony\\Contracts\\EventDispatcher\\Event",
         "Symfony\\Contracts\\EventDispatcher\\EventDispatcherInterface",
+        "Symfony\\Component\\EventDispatcher\\Event",
```
Comment (keradus) [inline .composer-require-checker.json]: https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/4665
